src/mumble/GlobalShortcut_win.cpp

```diff
@@ -130,576 +142,395 @@ GlobalShortcutWin::~GlobalShortcutWin() {
 }
 
 void GlobalShortcutWin::run() {
-	if (HAS_FAILED(DirectInput8Create(GetModuleHandle(nullptr), DIRECTINPUT_VERSION, IID_IDirectInput8,
-									  reinterpret_cast< void ** >(&pDI), nullptr))) {
-		qFatal("GlobalShortcutWin: Failed to create d8input");
-		return;
-	}
-
-	// Print the user's LowLevelHooksTimeout registry key for debugging purposes.
-	// On Windows 7 and greater, Windows will silently remove badly behaving hooks
-	// without telling the application. Users can tweak the timeout themselves
-	// with this registry key.
-	HKEY key    = nullptr;
-	DWORD type  = 0;
-	DWORD value = 0;
-	DWORD len   = sizeof(DWORD);
-	if (RegOpenKeyExA(HKEY_CURRENT_USER, "Control Panel\\Desktop", 0, KEY_READ, &key) == ERROR_SUCCESS) {
-		LONG err =
-			RegQueryValueExA(key, "LowLevelHooksTimeout", nullptr, &type, reinterpret_cast< LPBYTE >(&value), &len);
-		if (err == ERROR_SUCCESS && type == REG_DWORD) {
-			qWarning("GlobalShortcutWin: Found LowLevelHooksTimeout with value = 0x%lx",
-					 static_cast< unsigned long >(value));
-		} else if (err == ERROR_FILE_NOT_FOUND) {
-			qWarning("GlobalShortcutWin: No LowLevelHooksTimeout registry key found.");
-		} else {
-			qWarning(
-				"GlobalShortcutWin: Error looking up LowLevelHooksTimeout. (Error: 0x%lx, Type: 0x%lx, Value: 0x%lx)",
-				static_cast< unsigned long >(err), static_cast< unsigned long >(type),
-				static_cast< unsigned long >(value));
+#ifdef USE_XBOXINPUT
+	if (Global::get().s.bEnableXboxInput) {
+		m_xinput.reset(new XboxInput());
+		if (!m_xinput->isValid()) {
+			qWarning("GlobalShortcutWin: Failed to initialize XboxInput!");
+			m_xinput.reset();
 		}
 	}
-
-	// Wait for MainWindow's constructor to finish before we enumerate DirectInput devices.
-	// We need to do this because adding a new device requires a Window handle. (SetCooperativeLevel())
-	while (!Global::get().mw)
-		this->yieldCurrentThread();
-
+#endif
 #ifdef USE_GKEY
 	if (Global::get().s.bEnableGKey) {
-		gkey = new GKeyLibrary();
-		qWarning("GlobalShortcutWin: GKeys initialized, isValid: %d", gkey->isValid());
+		m_gkey.reset(new GKeyLibrary());
+		if (!m_gkey->isValid()) {
+			qWarning("GlobalShortcutWin: Failed to initialize GKey!");
+			m_gkey.reset();
+		}
 	}
 #endif
-
-#ifdef USE_XBOXINPUT
-	if (Global::get().s.bEnableXboxInput) {
-		xboxinput = new XboxInput();
-		ZeroMemory(&xboxinputLastPacket, sizeof(xboxinputLastPacket));
-		qWarning("GlobalShortcutWin: XboxInput initialized, isValid: %d", xboxinput->isValid());
+	// Wait for MainWindow's constructor to finish, so that we can get the window's handle.
+	while (!mumble_mw_hwnd) {
+		yieldCurrentThread();
```

Comment:
Is there no alternative to active waiting? :thinking:

Reply:
In theory we can wait for a signal, but ideally we should initialize `GlobalShortcut` after everything else.